lib/temporalio/testing/workflow_environment.rb

```diff
@@ -0,0 +1,95 @@
+require 'google/protobuf/well_known_types'
+
+module Temporalio
+  module Testing
+    # Workflow environment for testing workflows.
+    #
+    # Most developers will want to use the {Temporalio::Testing.start_time_skipping_environment} to
+    # start a test server process that automatically skips time as needed. Alternatively,
+    # {Temporalio::Testing.start_local_environment} may be used for a full, local Temporal server
+    # with more features.
+    #
+    # @note Unless using the above mentioned methods with an explicit block, you will need to call
+    # {#shutdown} after you're finished with this environment.
+    class WorkflowEnvironment
+      # @return [Temporalio::Connection] A connection to this environment.
+      attr_reader :connection
+
+      # @api private
+      def initialize(server, connection)
+        @server = server
+        @connection = connection
```

Comment:
Is this a namespace specific connection? I think most workflow test environments should provide a namespace-specific client. Arguably there's value in just accepting the client and and an _optional_ server in case they wanted to instantiate this abstraction from, say, a cloud client.

Also, you have to make sure there is a client interceptor here that enables automatic time skipping when the workflow result is being waited for by the user.

Reply:
Testing workflow environments are namespace specific though (at least Temporalite). Can you show an example of `start_local_environment` with a custom namespace and then making an API call on that? Am I expected to pass the namespace string to environment creation and to client creation, though the former may be a general test fixture and the latter may be each test?

What is the concern for giving people a (default) client to their environment?

Reply:
> Can you show an example of start_local_environment with a custom namespace and then making an API call on that?

```ruby
namespace = 'my-namespace'
Temporalio::Testing.start_local_environment(namespace: namespace) do |env|
  env.connection.workflow_service.start_workflow(...) # raw gRPC
  client = Temporalio::Client.new(env.connection, namespace)
  client.start_workflow(...) # high-level client
end
```

> Am I expected to pass the namespace string to environment creation and to client creation, though the former may be a general test fixture and the latter may be each test?

We can probably make namespace accessible via the `env.namespace` to make it a bit more convenient. This way you can also initialize it with a default namespace without providing one explicitly.

> Or is it because we force people to send a string namespace name to clients and separately force them to send a string namespace name to workers, that we accept forcing people to send yet another string namespace name to environment?

Yes, basically I've made everything explicit for now so that we can find the right abstraction to simplify the interface. But we definitely need one before the stable release, absolutely.

> What is the concern for giving people a (default) client to their environment?

We can implement `env.client` and return a default `Client` with the pre-configured namespace, no problem. The challenge is with the optional Client arguments (interceptors and converters) that you need to somehow pass. Doing this via the env initialization (e.g. `start_local_environment` similar to Python) feels wrong, because you're doing lots of extra work (and future coupling) for the benefit of hiding the namespace and injecting a time-skipping interceptor.

Maybe the answer here is some other component that abstracts the configuration (connection, namespace, interceptors, converter, identity, etc) and everything else then Client & Worker are initialized with it, without getting coupled to each other.

Reply:
> The challenge is with the optional Client arguments (interceptors and converters) that you need to somehow pass

Definitely would just be a default. What we've found very common for users is to shallow clone the client with just one or two options changed. This is why Python and .NET are providing such options that the client was created with - so they can be shallow cloned to, say, change one thing like adding an interceptor.

Reply:
Sure thing, I'll add a default client and create a ticket to provide shallow clone option for the `Client`

Reply:
:+1:

Also from above, you have to make sure there is a client interceptor here that enables automatic time skipping when the workflow result is being waited for by the user. (sorry to double up on suggestions in the same original message to this thread).